tunneler/src/main/kotlin/com/salesforce/pomerium/PomeriumTunneler.kt

```diff
@@ -173,23 +195,50 @@ class PomeriumTunneler(
                 withContext(NonCancellable) {
                     openTunnels.remove(route)
                     localServerSocket.close()
-                    selectorManager.close()
+                    lifetime.terminate()
+                    // Don't close selectorManager here since it is shared
                 }
             }
         }.invokeOnCompletion { e ->
+            // Remove the tunnel from openTunnels when the coroutine completes
             openTunnels.remove(route)
             localServerSocket.close()
-            selectorManager.close()
+            lifetime.terminate()
             if (e !is CancellationException) {
                 LOG.error("Unhandled exception in tunneling coroutine", e)
             }
         }
 
+        openTunnels[route] = disposable
+
+        lifetime.onTermination {
+            openTunnels.remove(route)
+            localServerSocket.close()
+            lifetime.terminate()
```

Comment:
Lifetime is already terminated (based on the onTermination callback). Does not look like we need to change the signature of the method and just use Lifetime instead.